### ISSUE TITLE
关于飞桨对异步计算的支持的请教

### ISSUE THREAD

**Issue by skywalk163** [PaddlePaddle/Paddle#28759]
最近在写动手学深度学习的飞桨版本，我主要负责第八章，这章里的第二节是异步计算。
http://zh.gluon.ai/chapter_computational-performance/async-computation.html
里面有这样一段：

      MXNet中的异步计算
        广义上讲，MXNet包括用户直接用来交互的前端和系统用来执行计算的后端。例如，用户可以使用不同的前端语言编写MXNet程序，如Python、R、Scala和C++。无论使用何种前端编程语言，MXNet程序的执行主要都发生在C++实现的后端。换句话说，用户写好的前端MXNet程序会传给后端执行计算。后端有自己的线程在队列中不断收集任务并执行它们。

        MXNet通过前端线程和后端线程的交互实现异步计算。异步计算指，前端线程无须等待当前指令从后端线程返回结果就继续执行后面的指令。为了便于解释，假设Python前端线程调用以下4条指令。

我想问一下，咱们飞桨有类似这样的设计概念吗？ 能否用1-2段话描述一下。 如果有装x的或者有助于理解学习的例子代码就更好了。

我自己在飞桨里了解到的异步方面，就是咱们的DataLoader异步加载这块，准备写上去一点。

**Comment by chenwhql** [comment]
![image](https://user-images.githubusercontent.com/22561442/99763747-84bae000-2b36-11eb-8f02-5a765450c80b.png)
这个例子，如果是CPU计算的话，也是打印的话才会计算完吗？这个有没有试过？

**Comment by chenwhql** [comment]
链接里面8.2.1和8.2.2的行为描述，在GPU环境下，飞桨也是一样的，需要print或者.numpy拿回值才是计算完的，python语句执行完的时候结果并没有计算完，但这不是由飞桨提供的并行能力，而是GPU CUDA设备本身的异步计算特性，CUDA计算是先拉起kernel将任务给GPU执行，然后CPU会继续执行后面的工作，调用wait_to_read函数、waitall函数、asnumpy函数、asscalar函数等，会触发CUDA的sync操作，从而等待结果算完再继续后面的工作

但对于CPU来讲，是没有异步操作的，语句执行完就是计算完了

**Comment by chenwhql** [comment]
如果MXNet在CPU上也是类似的异步，那实现机制上应该是不同的

**Comment by skywalk163** [comment]
谢谢，我理解下，就先按照这个意思处理。

**Comment by skywalk163** [comment]
飞桨默认（GPU和CPU通信）就是异步处理的， 那我们能像MXNet例子里面那样调用wait_to_read函数、waitall函数、asnumpy函数、asscalar函数，来进行同步处理，来对比代码运行时间吗？ 
我看了下，飞桨源代码里好像没有把这些函数再用API表达出来，顶多在c代码里比如NCCL的时候会用到waitall等函数进行同步。

**Comment by chenwhql** [comment]
您写的是动态图模式吧

动态图这四个API目前应该只有和`asnumpy`类似的`numpy`，可以用Tensor.numpy()达到同样的效果，其他几个暂时没有

**Comment by skywalk163** [comment]
是动态图，好的，我去试试。

**Comment by skywalk163** [comment]
我用这个程序测试了一下：

        %%writefile paddlemem.py
        import paddle 
        import os
        import sys
        import subprocess
        import time
        import paddle.nn as nn

        args = sys.argv

        train_dataset = paddle.vision.datasets.MNIST(mode='train')
        test_dataset = paddle.vision.datasets.MNIST(mode='test')
        lenet = paddle.vision.models.LeNet()

        # 加载训练集 batch_size 设为 64
        train_loader = paddle.io.DataLoader(train_dataset, batch_size=64, shuffle=True)

        # 监测内存
        def get_mem():
            res = subprocess.check_output(['ps', 'u', '-p', str(os.getpid())])
            return int(str(res).split()[15]) / 1e3

        def train():
            epochs = 1
            loss_tmp = paddle.to_tensor([0.2680041]) 
            adam = paddle.optimizer.Adam(learning_rate=0.001, parameters=lenet.parameters())
            # 用Adam作为优化函数
            dctime = time.time()
            for epoch in range(epochs):
                for batch_id, data in enumerate(train_loader()):
                    x_data, y_data = data
                    predicts = lenet(x_data)
                    loss = paddle.nn.functional.cross_entropy(predicts, y_data, reduction='mean')
                    acc = paddle.metric.accuracy(predicts, y_data, k=1)
                    avg_acc = paddle.mean(acc)
                    loss.backward()
                    if args[1] == 'sync':
                        loss.numpy # 同步函数
                    else:
                        loss_tmp.numpy 

                    if batch_id % 400 == 0:
                        # print(loss.numpy())
                        print("epoch: {}, batch_id: {}, loss is: {}, acc is: {}".format(epoch, batch_id, loss.numpy(), avg_acc.numpy()))
                    adam.step()
                    adam.clear_grad()
            endtime = time.time()-dctime
            print("time", endtime)
        # 启动train多进程训练，默认使用所有可见的GPU卡
        # import paddle.distributed as dist
        if __name__ == '__main__':
            mem =  get_mem()
            
            train()
            print('increased memory: %f MB' % (get_mem() - mem))
            print(get_mem(), mem )

分别执行 

        !python paddlemem.py sync
        !python paddlemem.py async

可以看到细微的差别，我想问一下，上面这种思路对吗 ？ 用一个无关变量进行 loss_tmp.numpy ，是否也会触发整段代码的同步 ？ 

或者就同步进行numpy操作，异步不进行操作 ？ 

            if args[1] == 'sync':
                loss.numpy() # 同步函数

结果同步竟然用的内存更多 

同步 

time 4.123139381408691
increased memory: 201.292000 MB

异步 

time 3.95578932762146
increased memory: 201.164000 MB

**Comment by skywalk163** [comment]
我把动手学深度学习的异步计算这一节写了一个[aistudio项目](https://aistudio.baidu.com/aistudio/projectdetail/196937) 共享出来。 

我想确认下，我这种使用numpy进行同步的方法是否正确？ 

我写一个经典的手写数字识别程序来做测试，用来测试同步函数和异步函数对内存的影响，顺便也对比下运行时间的不同。因为飞桨只有Tensor.numpy()函数和print函数来进行同步，所以这里用Tensor.numpy()进行同步操作，同时为了保证同步和异步同样的计算量，增加了一个loss_tmp变量，这样不管同步还是异步，都会执行Tensor.numpy()。 

也就是同步的时候，我用数字识别程序体里的loss变量来`loss.numpy()`同步， 在异步的时候，我是通过一个不参与主程序体的临时变量loss_tmp ，进行`loss_tmp.numpy()`来保证计算量相等，这样是否能如我所设想的那样不干扰异步操作 ？ 换句话说，是否程序体里任何一句tensor.numpy 会导致程序同步计算 ？

**Comment by chenwhql** [comment]
> 是否程序体里任何一句tensor.numpy 会导致程序同步计算

是这样的，这种写法不能区分同步和异步

**Comment by chenwhql** [comment]
任何一个var.numpy()，目前都会等待GPU的计算任务完成再返回结果

我不太确定异步计算这个事情是否是MXNet的卖点，但是训练过程中如果一般loss的值是需要拿回来观察训练的效果的，对于MXnet来讲也是把数据同步回来的

飞桨动态图目前不适合写这种整体的例子对比同步和异步的差别，或者比如你可以试下，10个batch调一次loss.numpy和每个batch都调用loss.numpy的差别

**Comment by skywalk163** [comment]
非常感谢，我按照你的建议重新写了例子，能看出差别来。 

我没有深入研究MXNet，好像它的异步计算算个卖点吧（主要是把几个CUDA同步函数都API出来，增加了python下的可控性啊）。
